modules/tribes/tests/client/components/TribesPage.component.test.js

```diff
@@ -0,0 +1,177 @@
+  label: 'Snails',
+}, {
+  _id: 'cccc',
+  id: 'cccc',
+  count: 1500,
+  slug: 'teddybears',
+  label: 'Teddy Bears',
+}];
+
+const dummyUser = {
+  _id: 'user0',
+  id: 'user0',
+  memberIds: ['cccc', 'aaaa'],
+};
+
+api.read.mockImplementation(async () => dummyTribes);
+api.join.mockImplementation(async id => ({ tribe: dummyTribes.find(tribe => tribe._id === id) }));
+api.leave.mockImplementation(async id => ({ tribe: dummyTribes.find(tribe => tribe._id === id) }));
+
+
+describe('TribesPage', () => {
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('not signed in', () => {
+
+    it('fetch tribes from api and show tribes on page', async () => {
+      // first no api should be called
+      expect(api.read).toHaveBeenCalledTimes(0);
+
+      const { getAllByRole } = render(<TribesPage onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // the tribes from api should be displayed
+      const tribes = getAllByRole('listitem');
+      expect(tribes).toHaveLength(dummyTribes.length + 1); // there is additionally the Suggest Tribe
+
+      tribes.slice(0, -1).forEach((tribe, i) => {
+        expect(tribe).toHaveTextContent(dummyTribes[i].label);
+      });
+
+      // during the test the api should be called only once
+      expect(api.read).toHaveBeenCalledTimes(1);
+    });
+
+    it('the join button should be a link to tribe page', async () => {
+      const { getAllByRole } = render(<TribesPage onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // join button
+      const tribes = getAllByRole('listitem').slice(0, -1);
+
+      tribes.forEach((tribe, i) => {
+        const button = rtl.getByText(tribe, 'Join');
+        expect(button).toHaveAttribute('href', `/signup?tribe=${dummyTribes[i].slug}`);
+      });
+    });
+
+  });
+
+  describe('signed in', () => {
+
+    it('show tribes on page', async () => {
+      // first no api should be called
+      expect(api.read).toHaveBeenCalledTimes(0);
+
+      const { getAllByRole } = render(<TribesPage user={dummyUser} onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // the tribes from api should be displayed
+      const tribes = getAllByRole('listitem');
+      expect(tribes).toHaveLength(dummyTribes.length + 1); // there is additionally the Suggest Tribe
+
+      tribes.slice(0, -1).forEach((tribe, i) => {
+        expect(tribe).toHaveTextContent(dummyTribes[i].label);
+      });
+
+      // during the test the api should be called only once
+      expect(api.read).toHaveBeenCalledTimes(1);
+    });
+
+    it('user is member of some tribes and not member of others', async () => {
+      const { getAllByText } = render(<TribesPage user={dummyUser} onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // the tribes from api should be displayed
+      const join = getAllByText('Join');
+      const leave = getAllByText('Joined');
+      expect(join).toHaveLength(1);
+      expect(leave).toHaveLength(2);
```

Comment:
I would tend to put the queries inline, and then you'll also see the linter suggest using the `query*` version, e.g.:

```js
expect(queryAllByText('Join')).toHaveLength(1);
expect(queryAllByText('Joined')).toHaveLength(2);
```

See https://testing-library.com/docs/dom-testing-library/api-queries for docs on get vs query.

Reply:
Fixed in 48216ee1f3691affb82eca49b6aaca06ed1271be.